README
Qore UUID module (uuid)

INTRODUCTION
------------
The Qore uuid module provides the ability to generate and parse UUIDs, Universally Unique IDentifiers
For more information about UUIDs, see:
    	http://en.wikipedia.org/wiki/Universally_Unique_Identifier

Basically this module provides a Qore-language API to libuuid on UNIX systems

see docs for more information


BUILDING
--------
Requires qore 0.8.0+ and libuuid to build and run

to configure the build, execute
   	configure --disable-debug

if the qore library cannot be found, then you can use:
   	configure --disable-debug --with-qore=<dir>

if libuuid cannot be found, then you can use:
   	configure --disable-debug --with-libuuid=<dir>

The qore binary also needs to be in the path so configure can determine the module directory

Then execute

make && make install

(or 'make && sudo make install' as needed)

please direct any questions to:
david@qore.org

or you can post in the qore forums at
        http://qore.org/forums


ChangeLog
2012-09-28 16:19  david_nichols

	* configure.ac: applied reini urban's multi-arch patches for debian
	  to all modules

2012-08-28 07:22  david_nichols

	* configure.ac, qore-uuid-module.spec: use bz2 instead of gz

2012-06-25 08:51  david_nichols

	* configure.ac: updated configure.ac logic to deal with /lib ->
	  /usr/lib and /lib64 -> /usr/lib64 as with Fedora 17 - the
	  previous logic was too clever for its own good

2012-05-24 17:16  david_nichols

	* Makefile.am, docs/doxygen-api-filter.q, qore-uuid-module.spec,
	  src/QC_UUID.qpp, src/uuid-module.h: updated build to use qdx;
	  fixed a bug in UUID::isNull() introduced with the qpp port, fixed
	  rpm building from the spec file

2012-05-23 15:54  david_nichols

	* Makefile.am, configure.ac, docs/doxyfile.tmpl,
	  docs/doxygen-api-filter.q, docs/footer_template.html,
	  docs/mainpage.doxygen.tmpl, docs/uuid-module-doc.html,
	  src/Makefile.am, src/QC_UUID.qpp,
	  src/single-compilation-unit.cpp, src/uuid-module.cpp,
	  src/uuid-module.h: updated the uuid module to qpp; set minimum
	  qore version to 0.8.5; deleted old stick wiki docs

2012-01-16 13:07  david_nichols

	* configure.ac: updated configure scripts to only set LIBSUFFIX=64
	  on 64-bit platfiorms if /usr/lib64 exists

2011-09-12 17:02  david_nichols

	* src/uuid-module.h: fix for buggy uuid headers (not const correct)

2011-09-12 16:26  david_nichols

	* configure.ac: added -mt to compile and link lines when building
	  on solaris with CC

2011-09-02 15:19  subik

	* ChangeLog: missing file for make dist

2011-08-22 16:10  david_nichols

	* RELEASE-NOTES, configure.ac, docs/uuid-module-doc.html,
	  src/uuid-module.h: updated the module to compile with the native
	  Windows (RPC) UUID API (no external libraries are necessary on
	  Windows anymore); updated the docs and release notes accordingly

2011-07-29 16:08  david_nichols

	* src/uuid-module.h: module is now working with ossp uuid and under
	  windows with mingw32

2011-07-29 15:59  david_nichols

	* README, configure.ac, src/Makefile.am, src/uuid-module.h: mingw32
	  changes; ported to ossp uuid API for mingw, which is different
	  from
	  the Linux/OSX/Solaris standard from Ted Ts'o, but still not
	  working

2010-12-17 17:00  david_nichols

	* README: separated xml functionality from qore and moved into
	  separate module ("xml") -
	  docs not yet moved

2010-11-08 16:47  david_nichols

	* docs/uuid-module-doc.html: fixed mysql and sybase auto-reconnect
	  logic to use the new Datasource::activeTransaction() API method

2010-10-29 08:44  david_nichols

	* docs/uuid-module-doc.html, src/uuid-module.cpp: updated docs

2010-10-29 08:37  david_nichols

	* RELEASE-NOTES, configure.ac, qore-uuid-module.spec,
	  src/uuid-module.h: updated to version 1.2; took the following
	  steps to ensure consistent behavior on all platforms:
	  * modified the default output (without any flags) to always be a
	  string with
	  lower-case hex digit letters
	  * updated implementation to manage case even if the uuid library
	  does not
	  support the uuid_unparse_lower() and uuid_unparse_upper()
	  functions

2010-06-30 10:56  david_nichols

	* docs/uuid-module-doc.html, src/uuid-module.cpp,
	  src/uuid-module.h: swapped argument order of UUID::get()

2010-06-30 10:53  david_nichols

	* README-BINARIES, RELEASE-NOTES, configure.ac,
	  docs/uuid-module-doc.html, qore-uuid-module.spec,
	  src/uuid-module.cpp, src/uuid-module.h, test/uuid-test.q: updated
	  release to 1.1
	  added new method: static UUID::get(int $gen_flags = UUID::None,
	  int $str_flags = UUID::None) returns string
	  updated test script
	  documented new method and made minor documentation corrections

2010-06-29 13:50  david_nichols

	* src/uuid-module.h, test/uuid-test.q: updated test script and
	  fixed compilation with g++

2010-06-29 13:37  david_nichols

	* test/uuid-test.q: updated test

2010-06-29 13:24  david_nichols

	* configure.ac, src/uuid-module.cpp, src/uuid-module.h,
	  test/uuid-test.q: solaris build fixes

2010-06-29 11:55  david_nichols

	* configure.ac, src/uuid-module.h: fixes for compiling on
	  darwin/OSX with uuid functions in the System library

2010-06-29 11:38  david_nichols

	* ., AUTHORS, COPYING, Makefile.am, README, RELEASE-NOTES,
	  configure.ac, docs, docs/uuid-module-doc.html, m4,
	  m4/acx_pthread.m4, qore-uuid-module.spec, reconf.sh, src,
	  src/Makefile.am, src/single-compilation-unit.cpp,
	  src/uuid-module.cpp, src/uuid-module.h, test, test/uuid-test.q:
	  checking in new uuid module



src/uuid-module.h
/* -*- mode: c++; indent-tabs-mode: nil -*- */
/*
  uuid-module.h
  
  Qore Programming Language

  Copyright 2010 - 2011 David Nichols

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

#ifndef _QORE_UUID_MODULE_H
#define _QORE_UUID_MODULE_H

#include "../config.h"

#include <qore/Qore.h>

#if (defined _WIN32 || defined __WIN32__) && ! defined __CYGWIN__
#define WIN_UUID 1
#else
#include <uuid.h>
#endif

#ifndef _UUID_STRING_T
typedef char* uuid_string_t;
#endif

// uuid flags
#define QUF_NONE        0
#define QUF_UPPER_CASE  1
#define QUF_LOWER_CASE  2
#define QUF_RANDOM      3
#define QUF_TIME        4
#define QUF_EMPTY       5

#ifdef WIN_UUID
typedef UUID q_uuid_t;
#else
#ifdef OSSP_UUID
typedef uuid_t* q_uuid_t;
#else
typedef uuid_t q_uuid_t;
#endif
#endif

#define Q_UUID_LEN 36

class QoreUUID : public AbstractPrivateData {
protected:
   q_uuid_t uuid;

   DLLLOCAL static QoreStringNode *getStringIntern(q_uuid_t &uuid, int flags = QUF_NONE) {
      if (flags == QUF_NONE)
         flags = QUF_LOWER_CASE;

#ifdef WIN_UUID
      UCHAR *StringUuid = 0;
#ifdef DEBUG
      RPC_STATUS rc =
#endif
      UuidToString(&uuid, &StringUuid);
#ifdef DEBUG
      assert(rc == RPC_S_OK);
#endif
      QoreStringNode *str = new QoreStringNode((const char *)StringUuid);
      RpcStringFree(&StringUuid);
      if (flags & QUF_UPPER_CASE)
         str->toupr();
      else
         str->tolwr();
#else // WIN_UUID
#ifdef OSSP_UUID
      size_t len = 0;
#ifdef DEBUG
      char *buf = 0;
      uuid_rc_t rc = uuid_export(uuid, UUID_FMT_STR, &buf, &len);
      if (rc != UUID_RC_OK)
         printd(0, "uuid_export error: %d: %s\n", rc, uuid_error(rc));
      assert(rc == UUID_RC_OK);
#else // !DEBUG
      uuid_export(uuid, UUID_FMT_STR, &buf, &len);
#endif // DEBUG
      QoreStringNode *str = new QoreStringNode(buf, len, len + 1, QCS_DEFAULT);
      if (flags & QUF_UPPER_CASE)
         str->toupr();
      else
         str->tolwr();
#else // OSSP_UUID
      QoreStringNode *str = new QoreStringNode;
      str->allocate(Q_UUID_LEN + 1);

#ifdef HAVE_UUID_UNPARSE_CASE
      if (flags & QUF_LOWER_CASE)
         uuid_unparse_lower(uuid, (char *)str->getBuffer());
      else
         uuid_unparse_upper(uuid, (char *)str->getBuffer());
      str->terminate(36);
#else
      uuid_unparse(uuid, (char *)str->getBuffer());
      str->terminate(36);
      if (flags & QUF_UPPER_CASE)
         str->toupr();
      else
         str->tolwr();
#endif // HAVE_UUID_UNPARSE_CASE
#endif // OSSP_UUID
#endif // WIN_UUID

      return str;
   }

   DLLLOCAL static void generateIntern(q_uuid_t &uuid, int flags = QUF_NONE) {
#ifdef WIN_UUID
      UuidCreate(&uuid);
#else
#ifdef OSSP_UUID
      if (flags & QUF_TIME)
         uuid_make(uuid, UUID_MAKE_V1);
      else if (!(flags & QUF_EMPTY))
         uuid_make(uuid, UUID_MAKE_V4);
#else
      if (flags & QUF_RANDOM)
         uuid_generate_random(uuid);
      else if (flags & QUF_TIME)
         uuid_generate_time(uuid);
      else if (flags & QUF_EMPTY)
         uuid_clear(uuid);
      else
         uuid_generate(uuid);
#endif
#endif
   }

public:
   DLLLOCAL QoreUUID(int flags = QUF_NONE) {
#ifdef OSSP_UUID
      uuid_create(&uuid);
#endif
      generateIntern(uuid, flags);
   }

   DLLLOCAL QoreUUID(const QoreString &uuid_str, ExceptionSink *xsink) {
#ifdef OSSP_UUID
      uuid_create(&uuid);
#endif
      set(uuid_str, xsink);
   }

   DLLLOCAL QoreUUID(const QoreUUID &old) {
#ifdef WIN_UUID
      // FIXME: test
      memcpy(&uuid, &old.uuid, sizeof(UUID));
#else
#ifdef OSSP_UUID
      uuid_create(&uuid);
      uuid_clone(old.uuid, &uuid);
#else
      uuid_copy(uuid, const_cast<QoreUUID &>(old).uuid);
#endif
#endif
   }

#ifdef OSSP_UUID
   DLLLOCAL ~QoreUUID() {
      uuid_destroy(uuid);
   }
#endif

   DLLLOCAL void generate(int flags = QUF_NONE) {
      generateIntern(uuid, flags);
   }

   DLLLOCAL QoreStringNode *getErrorString(ExceptionSink *xsink, const QoreString &uuid_str, const char *err = "cannot parse string") {
      QoreStringNode *desc = new QoreStringNode(err);
      desc->concat(": argument '");
      desc->concat(&uuid_str, xsink);
      if (desc->strlen() > 255) {
         desc->terminate(255);
         desc->concat("...");
      }
      desc->concat("' as a UUID string (must be in format like ex: '1b4e28ba-2fa1-11d2-883f-b9a761bde3fb')");
      return desc;
   }

   DLLLOCAL int set(const QoreString &uuid_str, ExceptionSink *xsink) {
#ifdef WIN_UUID
      RPC_STATUS rc = UuidFromString((unsigned char *)uuid_str.getBuffer(), &uuid);
      if (rc == RPC_S_OK)
         return 0;

      xsink->raiseException("UUID-STRING-ERROR", getErrorString(xsink, uuid_str));
      return -1;
#else
#ifdef OSSP_UUID
      uuid_rc_t rc = uuid_import(uuid, UUID_FMT_STR, uuid_str.getBuffer(), uuid_str.strlen());
      if (rc == UUID_RC_OK)
         return 0;

      xsink->raiseException("UUID-STRING-ERROR", getErrorString(xsink, uuid_str, uuid_error(rc)));
      return -1;
#else
      if (uuid_parse((char *)uuid_str.getBuffer(), uuid)) {
         xsink->raiseException("UUID-STRING-ERROR", getErrorString(xsink, uuid_str));
         return -1;
      }
      return 0;
#endif
#endif
   }

   DLLLOCAL QoreStringNode *toString(int flags = QUF_NONE) {
      return getStringIntern(uuid, flags);
   }

   DLLLOCAL bool isNull() const {
#ifdef WIN_UUID
      RPC_STATUS rc;
      UuidIsNil((UUID *)&uuid, &rc);
      return rc;
#else
#ifdef OSSP_UUID
      int result;
      uuid_isnil(uuid, &result);
      return result;
#else
      return (bool)uuid_is_null(const_cast<QoreUUID *>(this)->uuid);
#endif
#endif
   }

   DLLLOCAL void clear() {
#ifdef WIN_UUID
      UuidCreateNil(&uuid);
#else
#ifdef OSSP_UUID
      uuid_destroy(uuid);
      uuid_create(&uuid);
#else
      uuid_clear(uuid);
#endif
#endif
   }

   DLLLOCAL int compare(const QoreUUID &other) const {
#ifdef WIN_UUID
      RPC_STATUS rc;
      UuidCompare((UUID*)&uuid, (UUID*)&other.uuid, &rc);
      return rc;
#else
#ifdef OSSP_UUID
      int result;
      uuid_compare(uuid, other.uuid, &result);
      return result;
#else
      return uuid_compare(const_cast<QoreUUID*>(this)->uuid, const_cast<QoreUUID &>(other).uuid);
#endif
#endif
   }

   DLLLOCAL static QoreStringNode *get(int string_flags = QUF_NONE, int gen_flags = QUF_NONE) {
      q_uuid_t uuid;
#ifdef OSSP_UUID
      uuid_create(&uuid);
#endif
      generateIntern(uuid, gen_flags);
      return getStringIntern(uuid, string_flags);
   }
};

DLLLOCAL extern qore_classid_t CID_UUID;

#endif


src/uuid-module.cpp
/* indent-tabs-mode: nil -*- */
/*
  uuid Qore module

  Copyright (C) 2010 - 2012 David Nichols

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

#include "uuid-module.h"

static QoreStringNode *uuid_module_init();
static void uuid_module_ns_init(QoreNamespace *rns, QoreNamespace *qns);
static void uuid_module_delete();

// qore module symbols
DLLEXPORT char qore_module_name[] = "uuid";
DLLEXPORT char qore_module_version[] = PACKAGE_VERSION;
DLLEXPORT char qore_module_description[] = "uuid module";
DLLEXPORT char qore_module_author[] = "David Nichols";
DLLEXPORT char qore_module_url[] = "http://qore.org";
DLLEXPORT int qore_module_api_major = QORE_MODULE_API_MAJOR;
DLLEXPORT int qore_module_api_minor = QORE_MODULE_API_MINOR;
DLLEXPORT qore_module_init_t qore_module_init = uuid_module_init;
DLLEXPORT qore_module_ns_init_t qore_module_ns_init = uuid_module_ns_init;
DLLEXPORT qore_module_delete_t qore_module_delete = uuid_module_delete;
DLLEXPORT qore_license_t qore_module_license = QL_LGPL;

DLLLOCAL QoreClass* initUUIDClass(QoreNamespace &ns);

QoreNamespace UNS("UUID");

QoreStringNode *uuid_module_init() {
   // setup namespace
   QoreClass *QC_UUID = initUUIDClass(UNS);
   UNS.addSystemClass(QC_UUID);
   return 0;
}

void uuid_module_ns_init(QoreNamespace *rns, QoreNamespace *qns) {
   qns->addNamespace(UNS.copy());
}

void uuid_module_delete() {
}
